### ISSUE TITLE
How to get block and transactions data?

### ISSUE THREAD

**Issue by dimitriib** [bitcoinjs/bitcoinjs-lib#1562]
I trying to parse blk*.dat files via blkdat-stream for receiving some data. 
It's my code:

```js
glob('./blk/blk*.dat', {}, function (err, files) {
  console.log('files', files);
  const cs = CombinedStream.create();
  for (let j = 8; j < files.length; j++) {
    cs.append(fs.createReadStream(path.resolve(dataDir, files[j])))
  }
  cs.pipe(new BlockStream()).on('data', function (blockBuffer) {
    const block = BitcoinJSLib.Block.fromBuffer(blockBuffer);
    console.log('block', block);
    const ins = block.transactions[1].ins[1];
    const out = block.transactions[1].outs[1];
    const outAddress = BitcoinJSLib.address.fromOutputScript(out.script);
    let p2sh = BitcoinJSLib.payments.p2sh({
      witness: ins.witness,
      input: ins.script
    });
    const insAddress = p2sh.address;
  })
})
```

How I can get: 
hash and height from Block,
hash and fee from Transaction,
amount of tx and prev tx hash from Input,
and amount of tx from Output?

Thank you!

**Comment by junderw** [comment]
- hash and height from Block,

```js
const bitcoin = require('bitcoinjs-lib')
const block = bitcoin.Block.fromBuffer(blockBuffer)
const hash = block.getHash()
const blockId = block.getId() // hash byte order reversed as a hex string

// block height is only written in the block starting from height 227836
// see BIP34.
// for blocks less than that you will just have to count by connecting
// them together and counting from genesis block.
// height will be 3 bytes for another 142 years. then it will go to 4 bytes. the first byte of the script is length.
const heightBuff = block.transactions[0].ins[0].script.slice(1, 4)
const height = heightBuff[0] + (heightBuff[1] << 8) + (heightBuff[2] << 16)
```

- hash and fee from Transaction,
```js
const tx = bitcoin.Transaction.fromBuffer(txBuff)
const hash = tx.getHash()
const txId = tx.getId()

// Can't get fee without grabbing all the transactions referenced in the inputs
// assume a TX_CACHE where the key is the txId and the value is the Transaction object

const inputValue = tx.ins.reduce((total, input) => {
  const parentTx = TX_CACHE[Buffer.from(input.hash).reverse().toString('hex')]
  const spentOutput = parentTx.outs[input.index]
  return total + spentOutput.value
}, 0)
const outputValue = tx.outs.reduce((total, output) => {
  return total + output.value
}, 0)
const feeSatoshis = inputValue - outputValue
const feeRate = Math.ceil(feeSatoshis / tx.virtualSize()) // satoshi per byte
```

- amount of tx and prev tx hash from Input,
```
see above
```

- and amount of tx from Output?
```
Could you rephrase this? I don't understand.
```

**Comment by dimitriib** [comment]
@junderw many thanks, you are saved my day!
"and amount of tx from Output?" it's was question about outputValue and I got it from you.
Thank you again!

**Comment by dimitriib** [comment]
@junderw sorry, can you tell me more about TX_CACHE ?
Can I get it from blk*.dat files?

**Comment by junderw** [comment]
@dimitriib if you parse all the transactions in each block in order and every time you parse a transaction you enter the transaction into a TX_CACHE object, you can get it.

Node will run out of memory.

I suggest using a database of some sort instead of just a normal object.

**Comment by junderw** [comment]
You might also want to look at this:

https://github.com/bitcoinjs/fast-dat-parser

**Comment by dimitriib** [comment]
so TX_CACHE it's just array(for example) of all previous transactions?

**Comment by junderw** [comment]
key value store is easier, since all transactions have a unique ID (actually, there are two transactions in history with same hash and ID, which is why BIP34 came and required the height to be written in the block coinbase tx)

**Comment by dimitriib** [comment]
@junderw many thanks for your time!
Will research this question

**Comment by dimitriib** [comment]
@junderw 
I cant receive input address via script
`
 let p2sh = BitcoinJSLib.payments.p2sh({
      witness: ins.witness,
      input: ins.script
    });
`
I guess it's because it's not p2sh :-)
how I can get this sign type and receive addr? :-)

Sorry for disturbing again

**Comment by junderw** [comment]
> locktime

https://en.bitcoin.it/wiki/Protocol_documentation#tx

> how I can get transaction date and time?

You can't get that retroactively. The closest thing is the block timestamp. But remember that the block timestamp has a window of about 3 hours. So sometimes block x+1 has an earlier timestamp than block x... so you should not order blocks from timestamp.

In the blockchain, the only 100% accurate measurement of time is block order.

Also, you can deduce order of SOME txes within the same block by looking at the dependencies. ie. if tx A and tx B are both in the same block, and tx B input references tx A output, then you can deduce that tx A was broadcasted before tx B.

Other than that, timestamps are meaningless.

Most block explorers just take the block timestamp and use that as a "timestamp" for all the txes within the block.

> how I can get this sign type and receive addr?

The easiest way to get an address is to look at the output it references. The only real way to get the "type and address" is to parse the output itself. any input/witness based address generation should be considered "best guess" at best.

You could run all the payments against the output and see which one doesn't fail.

If more than one doesn't fail, you should throw an Error or mark it as "unknown address"...

I don't know what you're using this for. But showing the wrong address can cause people to send money to the wrong address if they plan on using it for checking addresses, so be careful.

**Comment by dimitriib** [comment]
@junderw no way to know which sign type for input address?

Now my solution is

`try {
      address = bitcoin.payments.p2sh({
        witness: witness,
        input: input,
      }).address;
    } catch (e) {
      err += e.message;
    }

    try {
      address = bitcoin.payments.p2pkh({
        witness: witness,
        input: input,
      }).address;
    } catch (e) {
      err += e.message;
    }

...
`

but it's too slow

**Comment by junderw** [comment]
"input address" does not exist. The way you are trying to do this is insecure.

Any block explorer that shows you an "input address" is really just looking at the OUTPUT that the input is pointing to and generating the address from the output script (which is easy and safe) because block explorers parse every transaction, so they will always have the output that is pointed at.

You should stop trying to do this.

If you want the "input address", go find the transaction the input is pointing to and look at the output.

Since blocks are not in the order of height, you will need to either order the blocks first, or make a second pass over all transactions after you've initially parsed each of them.

**Comment by junderw** [comment]
I would make a stream parser that just grabs the prevBlockHash from each header and gets the hash from each header and makes a note of the start byte position and filename. Then sort those, grab the byte position + filename tuples and go down them in order while parsing. That way you will always be guaranteed to have parsed the transaction that any input you parse is pointing to.

**Comment by junderw** [comment]
What database are you using to store your parsed data?

**Comment by dimitriib** [comment]
@junderw 
I tried using Postgresql, I used saving row by row, for inputs table, outputs table, blocks able, and transactions table. It's was pretty fast, around 2 sec for block. 
My next step is clustering this data, I wanna cluster all addresses which in one input, it's the reason why I need addresses for inputs.
PGSql to slow for this, cant do fast update, too much data. 
I gonna try MongoDB, and ClickHouse.
